### PR TITLE
Allow HcalPulseContainment timePhase definition as in SIM

### PR DESCRIPTION
**Pull request by mseidel42**
#### PR description:

We found that the containment correction is too large when used to correct the charge from a single time slice with default parameters. It turns out that the old algorithm has some additional shift (time0shift) and reversed behaviour of the timePhase parameter wrt simulation.
This PR allows for timePhase definition in HcalPulseContainment similar to SIM step, i.e., shift goes in the same direction so that parameters can be synchronized.

* So far, the change is only activated in the HcalPulseContainmentTest
  * NB: this falls back to the default pulse shape right now, as it cannot connect to the HcalDbService. Needs to be migrated to esConsumes? Tests were done with hard-coded reco shape 208, however
* It will be applied to PF1' scheme within HcaluLUTTPGCoder in a later PR
* Application in M0 and ZDC reco to be discussed

#### PR validation:

Containment correction for a single TS is now calculated within 6 - 31 ns of the pulse shape, similar to first TS in simulation.
The correction is the correct 1.56 instead of 2.18 for large pulses (see https://its.cern.ch/jira/browse/HCALDPG-18).

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34572	6ecc037
- code changes are technical for reco (to percolate a new argument in `HcalPulseContainment*` calls) in line with the PR description
- jenkins tests pass and comparisons with the baseline show no differences (the relval-input errors are unrelated)

**Comment by perrotta** [comment]
please test
(let retry and see if we can get rid of the unrelated relval-input errors)

File: CalibCalorimetry/HcalAlgos/plugins/HcalPulseContainmentTest.cc [hunk 1]
```diff
@@ -65,6 +65,9 @@ void HcalPulseContainmentTest::analyze(const edm::Event& iEvent, const edm::Even
   double corr4 = manager->correction(he1, 4, fixedphase_ns, fc);
   assert(corr4 == corr1);
   std::cout << corr1 << " " << corr2 << " " << corr3 << " " << corr4 << " " << std::endl;
+  // test 1TS correction at high energy
+  double corr5 = manager->correction(hb1, 1, fixedphase_ns, 100000.);
+  std::cout << corr5 << std::endl;
```
Comment (tvami) [inline CalibCalorimetry/HcalAlgos/plugins/HcalPulseContainmentTest.cc:70]: My understanding is that the use of std::cout is discouraged. Please use the relevant MessageLogger instead, https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger

File: CalibCalorimetry/HcalAlgos/src/HcalPulseContainmentAlgo.cc [hunk 2]
```diff
@@ -48,26 +54,30 @@ void HcalPulseContainmentAlgo::init(int num_samples) {
 #if 0
     char s[80];
     sprintf (s, "%7.3f %8.5f %8.5f\n", tmin, bin0val, bin1val);
-    cout << s;
+    std::cout << s << std::endl;
```
Comment (tvami) [inline CalibCalorimetry/HcalAlgos/src/HcalPulseContainmentAlgo.cc:57]: My understanding is that the use of std::cout is discouraged. Please use the relevant MessageLogger instead, https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger

File: CalibCalorimetry/HcalAlgos/src/HcalPulseContainmentAlgo.cc [hunk 3]
```diff
@@ -48,26 +54,30 @@ void HcalPulseContainmentAlgo::init(int num_samples) {
...
 
     if (bin1val > bin0val) {
       time0shiftns_ = shift_ns;
       break;
     }
   }
-
 #if 0
-  cout << "time0shiftns_ = " << time0shiftns_ << endl;
+  std::cout << "time0shiftns_ = " << time0shiftns_ << std::endl;
```
Comment (tvami) [inline CalibCalorimetry/HcalAlgos/src/HcalPulseContainmentAlgo.cc:66]: My understanding is that the use of std::cout is discouraged. Please use the relevant MessageLogger instead, https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger

File: CalibCalorimetry/HcalAlgos/src/HcalPulseContainmentAlgo.cc [hunk 4]
```diff
@@ -80,7 +90,7 @@ std::pair<double, double> HcalPulseContainmentAlgo::calcpair(double truefc) {
   char s[80];
   sprintf (s, "%8.2f %8.4f %8.4f %8.5f %8.5f %8.5f ",
 	   truefc, tmin, tmax, integral, corfactor, recofc);
-  cout << s;
+  std::cout << s << std::endl;
```
Comment (tvami) [inline CalibCalorimetry/HcalAlgos/src/HcalPulseContainmentAlgo.cc:93]: My understanding is that the use of std::cout is discouraged. Please use the relevant MessageLogger instead, https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger

**Review by tvami — COMMENTED** [review]
Comments about std::cout

**Comment by tvami** [comment]
> Comments about std::cout

Hi @mseidel42 , did you see these comments about std::cout? After that's fixed, I think we are happy to sign this from the AlCa point of view.

**Comment by mseidel42** [comment]
Hi @tvami, is it ok to use LogPrint?

**Comment by tvami** [comment]
> Hi @tvami, is it ok to use LogPrint?

Yes, that's fine!

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
+alca

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/34572	927d903 
- updates since the last reco signoff were in CalibCalorimetry/HcalAlgos (outside of reco)

**Comment by rekovic** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**